src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright 2019 The Polycube Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include <string>
+#include <sys/time.h>
+#include <iomanip>
+#include <iostream>
+#include <ctime>
+#include <fstream>
+#include <chrono>
+#include "unistd.h"
+#include "Packetcapture.h"
+#include "Packetcapture_dp_ingress.h"
+#include "Packetcapture_dp_egress.h"
+#define ON_T 0
+#define OFF_T 1
+#define BILLION 1000000000
+#define MILLION 1000000
+#define ONE_THOUSAND 1000
+
+
+typedef int bpf_int32; 
+typedef u_int bpf_u_int32;
+
+struct pcap_file_header {
+  bpf_u_int32 magic;      /* magic number */
+  u_short version_major;  /* major version number */
+  u_short version_minor;  /* minor version number */
+  bpf_int32 thiszone;     /* GMT to local correction */
+  bpf_u_int32 sigfigs;    /* accuracy of timestamps */
+  bpf_u_int32 snaplen;    /* max length of captured packets, in octets */
+  bpf_u_int32 linktype;   /* data link type */
+};
+
+struct pcap_pkthdr {
+  bpf_u_int32 ts_sec;
+  bpf_u_int32 ts_usec;  
+  bpf_u_int32 caplen;     /* number of octets of packet saved in file */
+  bpf_u_int32 len;        /* actual length of packet */
+};
+
+Packetcapture::Packetcapture(const std::string name, const PacketcaptureJsonObject &conf)
+  : TransparentCube(conf.getBase(), { packetcapture_code_ingress }, { packetcapture_code_egress }),
+    PacketcaptureBase(name) {
+  
+  /*
+   * The timestamp of the packet is acquired in the fast path using 'bpf_ktime_get_ns()'
+   * This function returns the time elapsed since system booted, in nanoseconds.
+   * See line 173 in Packetcapture_dp_egress.c and Packetcapture_dp_ingress.c
+   * 
+   * The packet's timestamp must be in epoch so now we will store the current
+   * time in epoch format and then the packet capture time offset,that was stored in the fast path,
+   * will be added to it.
+   * 
+   * Now we are calculating the system start time in epoch.
+   * 'timeval struct ts' rapresents the system start time in epoch calculated as
+   * actual time - system uptime
+   * 
+   * See line 138 of this file for more details
+   */
+
+
+  gettimeofday(&ts, NULL);              //getting actual time
+  std::uint64_t uptime = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
```

Comment:
@DavideAG I might be wrong but, what is the different between `gettimeofday` and uptime?
It looks to me that they are just two different ways of getting the current time but the produced result should be the same.
Moreover, if I remember correctly, you want to subtract the uptime from the current time and then add the value that you get from the data path to retrieve the final time on which the packet was captured.
If I am not wrong, the way you are calculating the uptime is not correct (see [1]). Am I missing something?

[1] https://stackoverflow.com/questions/30095439/how-do-i-get-system-up-time-in-milliseconds-in-c

Reply:
Thank you for your comment @sebymiano !

I changed this part of the code and it will be fixed in the next commit.